### DISCUSSION TITLE
Modifications only enabled for specific input source

### DISCUSSION THREAD

**Discussion by bykiandra** [pqrs-org/Karabiner-Elements#3670] [Q&A]
Hi all,

Only just started using Karabiner as I have a specific problem that needs solving.

So, I've been learning Japanese for a little while now, and I just got myself some new keycaps with the Hiragana characters to make learning on the computer easier. However, some of the characters on the keyboard don't match the characters for the input source, namely the equals sign, backslash and the closing curly bracket that have been shuffled around.

Now, for most people that might not be too much of a problem to handle when using the keyboard in English, as those characters aren't exactly used on a regular basis. For me however, I'm a developer, so I'm using those characters A LOT.

Is there some way I can set the changes to only be enabled for a specific input source?

**Comment by peterjc** [comment]
You can restrict your rules to only run on certain devices - would that work for you? I use this approach to remap my external keyboard but not the laptop's own keyboard.

Under the GUI settings, "Devices" from left hand pane, find the relevant keyboard(s) and you can toggle modify events on or off.

**Reply by peterjc** [reply]
Here's an example from my setup where I have a Japanese laptop keyboard (where I have told KE not to apply any rules) with an external UK layout Microsoft Ergonomic Keyboard 4000. I have left macOS setup with the Japanese layout (which Japanese chacters aside is almost a USA layout), which means typing on the UK keyboard I don't get punctuation matching the labels on the keys).

I therefore have 9 rules to fix this, all restricted to match the USB identifiers of that keyboard (just in case I plug in another keyboard - for example while travelling), using:

```json
            "conditions": [
                {
                    "identifiers": [
                        { "vendor_id": 1118 },
                        { "product_id": 219 }
                    ],
                    "type": "device_if"
                }
            ],
```

You can get those USB IDs in several way, most easily via the device list in KE itself.

Complete example:

```json
{
    "description": "UK style £ pound from Shift+3, rather than US/JP # pound",
    "manipulators": [
        {
            "conditions": [
                {
                    "identifiers": [
                        { "vendor_id": 1118 },
                        { "product_id": 219 }
                    ],
                    "type": "device_if"
                }
            ],
            "from": {
                "key_code": "3",
                "modifiers": { "mandatory": ["shift"] }
            },
            "to": [
                {
                    "key_code": "3",
                    "modifiers": ["option"],
                    "repeat": true
                }
            ],
            "type": "basic"
        }
    ]
}
}
```

**Comment by tekezo** [comment]
You can use `input_source_if` for this purpose.
https://karabiner-elements.pqrs.org/docs/json/complex-modifications-manipulator-definition/conditions/input-source/
